Closes #1869 Alphabetical filter and search view for person

BaoNguyen09 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Add a customized Glossary view using alphabetical index
## Description
<!--- Describe your changes in detail (include keywords close/fix/resolve) -->
- Create a glossary view that builds its glossary from the alphabetical index letter field rather than the title
- Use  `hook_views_post_execute` to manually add letter rows to the glossary attachment which are missing from the glossary because they have no results
- Make the view of glossory attachment the same with az_alphabetical_listing by overriding a template
- Add groupings for all person nodes
- Add an exposed search filter

### Release notes
<!--- Delete this line (and the closing comment line) to enable release notes.

If this change requires release notes: provide a summary of changes, how to use this change, and any related links. This content will be pasted in the [release notes](https://github.com/az-digital/az_quickstart/releases). Use markdown format to ensure proper pasting of information. [Release notes example](https://github.com/az-digital/az_quickstart/releases/tag/2.11.0)

Make sure to add the `release notes` label to this PR.

```
Add markdown of release notes here.
```

Delete this line to enable release notes.  -->


## Related issues
Closes #1869 
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## How to test
<!--- Please describe in detail how reviewers can test your changes -->
<!--- Include details of your testing environment and the tests you ran -->
1. Enable `field_az_alphabetical_index` at `/admin/structure/types/manage/az_person/fields`
2. For each person node, make sure this field is populated (if field is enabled before the node is created, it will be prepopulated)
3. Enable `Quickstart Alphabetical Listing` module in `/admin/modules`
4. Now it's all set up, go to `/alphabetical-glossary` to check, you should see: an alphabetical filter, a search bar, and all the nodes are groups based on their alphabetical index

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

### Arizona Quickstart (install profile, custom modules, custom theme)
- **Patch release changes**
   - [ ] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
   - [ ] Adding experimental module
   - [ ] Update experimental module
- **Minor release changes**
   - [ ] New feature
   - [x] Breaking or visual change to existing behavior
   - [ ] Upgrade experimental module to stable
   - [ ] Enable existing module by default or database update
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- **Other or unknown**
   - [ ] Other or unknown

### Drupal core
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major or minor level update
- **Other or unknown**
   - [ ] Other or unknown

### Drupal contrib projects
- **Patch release changes**
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- **Minor release changes**
   - [ ] Major level update
- **Other or unknown**
   - [ ] Other or unknown

## Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] My change requires release notes.

BaoNguyen09 committed:
feat: add a draft for alphabetical glossary view

BaoNguyen09 committed:
Merge remote-tracking branch 'origin/main' into 1869-alphabetical-filter-and-search-view-for-person

BaoNguyen09 committed:
feat: add a hook to implement all letters after the view is executed

BaoNguyen09 committed:
fix: add a forgotten field and remove one unused block

BaoNguyen09 committed:
fix: fix bugs+errors found by phpcs and phpstan

BaoNguyen09 committed:
Merge branch 'main' into 1869-alphabetical-filter-and-search-view-for-person

BaoNguyen09 committed:
chore: shorten comments to < 80 characters

BaoNguyen09 committed:
Merge remote-tracking branch 'origin/1869-alphabetical-filter-and-search-view-for-person' into 1869-alphabetical-filter-and-search-view-for-person

BaoNguyen09 committed:
Merge branch 'main' into 1869-alphabetical-filter-and-search-view-for-person

BaoNguyen09 committed:
feat: override theme template of glossary summary

BaoNguyen09 committed:
Merge remote-tracking branch 'origin/main' into 1869-alphabetical-filter-and-search-view-for-person

BaoNguyen09 committed:
Merge remote-tracking branch 'origin/1869-alphabetical-filter-and-search-view-for-person' into 1869-alphabetical-filter-and-search-view-for-person

tadean commented on modules/custom/az_alphabetical_listing/az_alphabetical_listing.module:
```diff
@@ -41,3 +104,86 @@ function az_alphabetical_listing_theme_registry_alter(&$theme_registry) {
+ * Implements hook_views_post_execute().
+ */
+function az_alphabetical_listing_views_post_execute(ViewExecutable $view) {
+  $entity_type_id = $view->getBaseEntityType()->id();
```
This seems like it should be moved inside the if statement so that it does not run for every view.

tadean commented on modules/custom/az_alphabetical_listing/az_alphabetical_listing.module:
```diff
@@ -27,7 +90,7 @@ function az_alphabetical_listing_theme_registry_alter(&$theme_registry) {
     if (!isset($theme_registry[$key]['theme path']) || strpos($theme_registry[$key]['theme path'], 'themes/') === FALSE) {
       // Copy default node registry entry for node entries.
-      if (strpos($key, 'node') === 0) {
+      if (strpos($key, 'node') === 0 && ($key !== 'views_view_summary__az_alphabetical_glossary_attachment')) {
```
I think this check should happen for the `if (strpos($key, 'views') === 0) {` line, rather than the node-based if statement

tadean commented on modules/custom/az_alphabetical_listing/az_alphabetical_listing.module:
```diff
@@ -15,6 +20,64 @@ function az_alphabetical_listing_preprocess_views_view(&$variables) {
+    // Create a quick lookup map from $view->result for easier access
+    // Keyed by the letter, value is the #existed_in_original flag.
+    $original_data_map = [];
+    foreach ($view->result as $result_row_from_post_execute) {
+      if (isset($result_row_from_post_execute->{$views_field_alias_for_letter})) {
+        $letter = $result_row_from_post_execute->{$views_field_alias_for_letter};
+        $existed = FALSE;
+        if (isset($result_row_from_post_execute->{'#existed_in_original'})) {
+          $existed = (bool) $result_row_from_post_execute->{'#existed_in_original'};
+        }
+        $original_data_map[$letter] = $existed;
+      }
+    }
```
This loop seems like it can be avoided by usage of the `num_records` property present on the original rows - this essentially calculates the same information.

BaoNguyen09 committed:
Merge branch 'main' into 1869-alphabetical-filter-and-search-view-for-person

BaoNguyen09 committed:
update: view attachment is switched to list-style

BaoNguyen09 committed:
fix: some logics + bugs based on troy review

BaoNguyen09 committed:
Merge remote-tracking branch 'origin/1869-alphabetical-filter-and-search-view-for-person' into 1869-alphabetical-filter-and-search-view-for-person

BaoNguyen09 committed:
fix: correct style of for glossary attachment

i update the template to conditionally include the 'disabled' css class for letters without records
i add this 'disabled' field into the row variable so that template can use it

tadean commented on modules/custom/az_alphabetical_listing/az_alphabetical_listing.module:
```diff
@@ -41,3 +85,90 @@ function az_alphabetical_listing_theme_registry_alter(&$theme_registry) {
+      if (isset($original_row->_entity)) {
+        /** @var \Drupal\Core\Entity\ContentEntityInterface $entity */
+        $entity = $original_row->_entity;
+        $field_items = $entity->get('field_az_alphabetical_index');
+        if (!$field_items->isEmpty()) {
+          // Use the actual field value (key) not the label.
+          $letters_in_original_results[$field_items->value] = $original_row->num_records;
+        }
```
Since this field may or may not exist on a given content type, it seems necessary to check for its existence. A php error seems to result for this view if the field has not been added for the content type.
```suggestion
        if ($entity->hasField('field_az_alphabetical_index')) {
          $field_items = $entity->get('field_az_alphabetical_index');
          if (!$field_items->isEmpty()) {
            // Use the actual field value (key) not the label.
            $letters_in_original_results[$field_items->value] = $original_row->num_records;
          }
        }
```

BaoNguyen09 committed:
Update modules/custom/az_alphabetical_listing/az_alphabetical_listing.module

Co-authored-by: tadean <51835841+tadean@users.noreply.github.com>

BaoNguyen09 committed:
Merge branch 'main' into 1869-alphabetical-filter-and-search-view-for-person

BaoNguyen09 committed:
feat: enable grouping+update display machine name

BaoNguyen09 committed:
feat: add an "all" filter button

BaoNguyen09 committed:
Merge remote-tracking branch 'origin/1869-alphabetical-filter-and-search-view-for-person' into 1869-alphabetical-filter-and-search-view-for-person

tadean commented on modules/custom/az_alphabetical_listing/config/install/views.view.az_alphabetical_glossary.yml:
```diff
@@ -0,0 +1,514 @@
+    - node
+    - options
+    - user
+id: alphabetical_glossary
```
```suggestion
id: az_alphabetical_glossary
```

BaoNguyen09 committed:
Update modules/custom/az_alphabetical_listing/config/install/views.view.az_alphabetical_glossary.yml

Co-authored-by: tadean <51835841+tadean@users.noreply.github.com>

BaoNguyen09 committed:
Update modules/custom/az_alphabetical_listing/config/install/views.view.az_alphabetical_glossary.yml

Co-authored-by: tadean <51835841+tadean@users.noreply.github.com>

BaoNguyen09 committed:
Update modules/custom/az_alphabetical_listing/config/install/views.view.az_alphabetical_glossary.yml

Co-authored-by: tadean <51835841+tadean@users.noreply.github.com>

BaoNguyen09 committed:
Update modules/custom/az_alphabetical_listing/config/install/views.view.az_alphabetical_glossary.yml

Co-authored-by: tadean <51835841+tadean@users.noreply.github.com>

BaoNguyen09 committed:
Update modules/custom/az_alphabetical_listing/config/install/views.view.az_alphabetical_glossary.yml

Co-authored-by: joshuasosa <4400438+joshuasosa@users.noreply.github.com>

BaoNguyen09 committed:
Merge branch 'main' into 1869-alphabetical-filter-and-search-view-for-person

BaoNguyen09 committed:
Merge branch 'main' into 1869-alphabetical-filter-and-search-view-for-person

BaoNguyen09 committed:
fix: export the right grouping config

BaoNguyen09 committed:
Merge branch 'main' into 1869-alphabetical-filter-and-search-view-for-person

BaoNguyen09 committed:
feat: add an exposed filter to search name

BaoNguyen09 committed:
Merge branch 'main' into 1869-alphabetical-filter-and-search-view-for-person

BaoNguyen09 committed:
Merge branch 'main' into 1869-alphabetical-filter-and-search-view-for-person

BaoNguyen09 committed:
add a draft block for this view

BaoNguyen09 committed:
refactor: remove the page display for this view

BaoNguyen09 committed:
Merge branch 'main' into 1869-alphabetical-filter-and-search-view-for-person

BaoNguyen09 committed:
Merge branch '1869-alphabetical-filter-and-search-view-for-person' of github.com:az-digital/az_quickstart into 1869-alphabetical-filter-and-search-view-for-person
